File: 0118-pascals-triangle/0118-pascals-triangle.cpp
class Solution
{
public:
    vector<vector<int> > generate(int numRows)
    {
        vector<vector<int> > ret;
        ret.reserve(numRows);
        for (int n = 0; n < numRows; n++)
        {
            vector<int> row;
            row.push_back(1); 
            row.reserve(n + 1);
            for (int k = 1; k <= n; k++)
            {
                row.push_back(row.back() * (n + 1 - k) / k);
            }
            ret.push_back(row);
        }
        return ret;
    }
};

File: 0118-pascals-triangle/0118-pascals-triangle.py
# time complexity: O(n^2)
# space complexity: O(n^2)
from typing import List


# class Solution:
#     def generate(self, numRows: int) -> List[List[int]]:
#         res = []
#         for i in range(numRows):
#             row = [1]
#             for k in range(1, i+1):
#                 row.append(row[-1] * (i + 1 - k) // k)
#             res.append(row)
#         return res

class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        dp = [[0 for _ in range(j)] for j in range(1, numRows+1)]
        for i in range(numRows):
            dp[i][0] = 1
            dp[i][i] = 1
        for i in range(2, numRows):
            for j in range(1, i):
                dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
        return dp


print(Solution().generate(5))
